drivers/adc/adm1177/adm1177.c

```diff
@@ -0,0 +1,274 @@
+	status = no_os_i2c_init(&dev->i2c_desc, &init_param->i2c_init);
+	if(status) {
+		no_os_free(dev);
+		return status;
+	}
+
+	*device = dev;
+
+	return status;
+}
+
+/**************************************************************************//**
+ * @brief Free the resources allocated by adm1177_init().
+ *
+ * @param device - The device structure.
+ *
+ * @return 0 in case of success, negative error code otherwise.
+******************************************************************************/
+int adm1177_remove(struct adm1177_dev *device)
+{
+	int ret;
+
+	ret = no_os_i2c_remove(device->i2c_desc);
+
+	no_os_free(device);
+
+	device = NULL;
+
+	return ret;
+}
+
+/**************************************************************************//**
+ * @brief - In the write command byte operation, the master device sends a
+ *          command byte to the slave device.
+ *
+ * @param device - The device structure.
+ *
+ * @param cmd - The command byte indicates which operation to be performed.
+ *
+ * @param reg_address - In case of a write extended command byte operation
+ * 			one of the ADM1177_REG_ALERT_EN, ADM1177_REG_ALERT_TH
+ * 			or ADM1177_REG_CONTROL will be used.
+ * 			In case of a write command byte operation reg_address
+ * 			will be set to 0.
+ *
+ * @return 0 in case of success, -1 otherwise.
+******************************************************************************/
+int adm1177_write(struct adm1177_dev *device, uint8_t cmd, uint8_t reg_address)
+{
+	if(reg_address & ADM1177_EXT_CMD_MASK) {
+		uint8_t data[2] = {reg_address, cmd};
+
+		return no_os_i2c_write(device->i2c_desc, data, 2, 1);
+	} else {
+		switch(cmd) {
+		case ADM1177_CMD_V_CONT:
+		case ADM1177_CMD_V_ONCE:
+			device->chan_en = ADM1177_VOLTAGE_EN;
+			break;
+		case ADM1177_CMD_I_CONT:
+		case ADM1177_CMD_I_ONCE:
+			device->chan_en = ADM1177_CURRENT_EN;
+			break;
+		case (ADM1177_CMD_V_CONT | ADM1177_CMD_I_CONT):
+		case (ADM1177_CMD_V_ONCE | ADM1177_CMD_I_ONCE):
+			device->chan_en = ADM1177_VOLTAGE_AND_CURRENT_EN;
+			break;
+		default:
+			return -EINVAL;
+		}
+		return no_os_i2c_write(device->i2c_desc, &cmd, 1, 1);
+	}
+}
+
+/**************************************************************************//**
+ * @brief - Gets the voltage/current readback from the ADM1177s I2C ADC,
+ *	    if the device is initialized right and transmits the requested
+ *	    values.
+ *
+ * @param device - The device structure.
+ *
+ * @param conv_voltage - The variable through which the voltage read from the
+ * 			 ADC is received if requested.
+ *
+ * @param conv_current - The variable through which the current read from the
+ * 			 ADC is received if requested.
+ *
+ * @return 0 in case of success, -1 otherwise.
+******************************************************************************/
+int adm1177_readback(struct adm1177_dev *device,
+		     uint16_t* conv_voltage,
+		     uint16_t* conv_current)
+{
+	int ret;
+
+	uint8_t data[3];
+
+	switch(device->chan_en) {
```

Comment:
i'd approach this readback function similar to https://github.com/analogdevicesinc/linux/blob/9e1304f0416a5bd3c79c48bc73d24bf0978a47d7/drivers/iio/adc/adm1177.c#L92-L103

I'd readback only one value based on the user request (either voltage or current).

The current approach might introduce erroneous read-backs in the upper layers. For case `ADM1177_CURRENT_EN ` the content of `conv_voltage` might be a older reading/ garbage value (but without any warning to the user) same for case `ADM1177_VOLTAGE_EN` and `conv_current`.

Reply:
I understand what you're saying, but I'm not really sure how I can approach it similar to the linux driver, in the way that if i write a command to the ADM saying I only need to read current then the current data comes in a different form, same for the voltage. So in my opinion I need that third case. 
What I can do is for the Voltage only readback and Current only readback, to make the other one 0 and mention it in the documentation, this way no erroneous read-backs will appear in the upper layers.

Reply:
Or you can add some wrappers. Rename this function `adm1177_readback_raw` and add some wrappers where you interpret the data `adm1177_read_voltage_raw` `adm1177_read_current_raw` `adm1177_read_voltage_and_current_raw`. Just a thought.